src/nbl/video/device_limits.json

```diff
@@ -0,0 +1,2751 @@
+                    "Vulkan 1.4 non-optional requires but poor support"
+                ]
+            },
+            {
+                "type": "bool",
+                "name": "shaderZeroInitializeWorkgroupMemory",
+                "value": false,
+                "comment": "or VK_KHR_zero_initialize_workgroup_memory"
+            }
+        ],
+        "nabla": [
+            {
+                "groupComment": [
+                    "Nabla"
+                ]
+            },
+            {
+                "type": "uint32_t",
+                "name": "computeUnits",
+                "value": 0
+            },
+            {
+                "type": "bool",
+                "name": "dispatchBase",
+                "value": false,
+                "comment": "true in Vk, false in GL"
+            },
+            {
+                "type": "bool",
+                "name": "allowCommandBufferQueryCopies",
+                "value": false
+            },
+            {
+                "type": "uint32_t",
+                "name": "maxOptimallyResidentWorkgroupInvocations",
+                "value": 0,
+                "comment": "its 1D because multidimensional workgroups are an illusion"
+            },
+            {
+                "type": "uint32_t",
+                "name": "maxResidentInvocations",
+                "value": 0,
+                "comment": "These are maximum number of invocations you could expect to execute simultaneously on this device"
+            },
+            {
+                "type": "asset::CGLSLCompiler::E_SPIRV_VERSION",
+                "name": "spirvVersion",
+                "value": "asset::CGLSLCompiler::E_SPIRV_VERSION::ESV_1_8"
+            },
+            {
+                "type": "function",
+                "function": [
+                    "// utility functions",
+                    "// In the cases where the workgroups synchronise with each other such as work DAGs (i.e. `CScanner`),",
+                    "// `workgroupSpinningProtection` is meant to protect against launching a dispatch so wide that",
+                    "// a workgroup of the next cut of the DAG spins for an extended time to wait on a workgroup from a previous one.",
+                    "inline uint32_t computeOptimalPersistentWorkgroupDispatchSize(const uint64_t elementCount, const uint32_t workgroupSize, const uint32_t workgroupSpinningProtection=1u) const",
+                    "{",
+                    "    assert(elementCount!=0ull && \"Input element count can't be 0!\");",
+                    "    const uint64_t infinitelyWideDeviceWGCount = (elementCount-1ull)/(static_cast<uint64_t>(workgroupSize)*static_cast<uint64_t>(workgroupSpinningProtection))+1ull;",
+                    "    const uint32_t maxResidentWorkgroups = maxResidentInvocations/workgroupSize;",
+                    "    return static_cast<uint32_t>(core::min<uint64_t>(infinitelyWideDeviceWGCount,maxResidentWorkgroups));",
+                    "}"
```

Comment:
if you follow the
```cpp
Some handwritten code

struct MyStruct
{
#include "python_generated.h"

More handwritten code
};

Some more handwritten code
```

pattern, you can add some handwritten members/methods to the struct.

Reply:
Done